gadgets/deadlock/go/program.go

```diff
@@ -0,0 +1,254 @@
+		api.Warnf("failed to add tag to field: %s", err)
+		return 1
+	}
+
+	// Build the graph for deadlock detection using edges from the mutex datasource
+	dsMutex.SubscribeArray(func(source api.DataSource, data api.DataArray) error {
+		affectedPIDs := make(map[uint32]struct{})
+		for i := 0; i < data.Len(); i++ {
+			datum := data.Get(i)
+			mutex1, err := mutex1F.Uint64(datum)
+			if err != nil {
+				api.Warnf("failed to get mutex1: %s", err)
+				continue
+			}
+			mutex2, err := mutex2F.Uint64(datum)
+			if err != nil {
+				api.Warnf("failed to get mutex2: %s", err)
+				continue
+			}
+			tid, err := tidF.Uint32(datum)
+			if err != nil {
+				api.Warnf("failed to get tid: %s", err)
+				continue
+			}
+			pid, err := pidF.Uint32(datum)
+			if err != nil {
+				api.Warnf("failed to get pid: %s", err)
+				continue
+			}
+			mutex1Stack, err := mutex1StackF.Uint64(datum)
+			if err != nil {
+				api.Warnf("failed to get mutex1_stack_id: %s", err)
+				continue
+			}
+			mutex2Stack, err := mutex2StackF.Uint64(datum)
+			if err != nil {
+				api.Warnf("failed to get mutex2_stack_id: %s", err)
+				continue
+			}
+			comm, err := commF.String(datum)
+			if err != nil {
+				api.Warnf("failed to get comm: %s", err)
+				continue
+			}
+			mntnsId, err := mntnsIdF.Uint64(datum)
+			if err != nil {
+				api.Warnf("failed to get mntns_id: %s", err)
+				continue
+			}
+
+			// Record process information for the PID
+			if _, exists := pidInfo[pid]; !exists {
+				pidInfo[pid] = ProcInfo{
+					comm:    comm,
+					mntnsId: mntnsId,
+				}
+			}
+			// Update the graph for the PID
+			g, exists := graphs[pid]
+			if !exists {
+				g = digraph.NewDiGraph()
+				graphs[pid] = g
+			}
+			g.AddEdge(mutex1, mutex2, map[string]interface{}{
+				"tid":           tid,
+				"mutex1StackId": mutex1Stack,
+				"mutex2StackId": mutex2Stack,
+			})
+			affectedPIDs[pid] = struct{}{}
+		}
+
+		// Check for cycles only in the affected PIDs
+		for pid := range affectedPIDs {
+			g := graphs[pid]
+			cycles := digraph.FindCycles(g)
+			if _, exists := detectedCycles[pid]; !exists {
+				// Create entry for the PID
+				detectedCycles[pid] = make(map[string]struct{})
+			}
+			if len(cycles) == 0 || len(cycles) == len(detectedCycles[pid]) {
+				continue // no new cycles detected
+			}
+			for _, cycle := range cycles {
+				cycleKey := cycleToKeyString(cycle)
+				if _, exists := detectedCycles[pid][cycleKey]; !exists {
+					// New cycle detected
+					detectedCycles[pid][cycleKey] = struct{}{}
+
+					// Emit the deadlock packet for the new cycle
+					packet, err := dsDeadlock.NewPacketSingle()
+					if err != nil {
+						api.Warnf("failed to create new packet: %s", err)
+						continue
+					}
+					deadlockPidF.SetUint32(api.Data(packet), pid)
+					deadlockNodesF.SetUint64(api.Data(packet), uint64(len(cycle)))
+					deadlockCommF.SetString(api.Data(packet), pidInfo[pid].comm)
+					deadlockMntnsIdF.SetUint64(api.Data(packet), pidInfo[pid].mntnsId)
```

Comment:
Earlier we were storing all attributes common for a PID with every node in the graph (earlier we were taking the mntns id from the first node of the entire cycle only, when creating a packet), meaning lots of unnecessary data duplication. Now, I have separated the attributes associated with a PID into a separate map. 

This also makes the code more readable and sensible (imo) as attributes related to the process are separated from attributes specific to an edge, instead of stuffing PID specific attributes into each edge in a PID's graph like we were doing before.

Reply:
Pushed another small optimization. This if of course under the assumption that a PID will have one process name and only one mntns ID of it's source.